src/index.ts

```diff
@@ -2,56 +2,87 @@ import { join } from 'path'
   let viteConfig: ResolvedConfig
   let options: ResolvedVitePWAOptions
+  let useRegister = false
 
-  return {
-    name: 'vite-plugin-pwa',
-    enforce: 'post',
-    apply: 'build',
-    configResolved(config) {
-      viteConfig = config
-      options = resolveOptions(userOptions, viteConfig)
-    },
-    transformIndexHtml: {
+  return [
+    {
+      name: 'vite-plugin-pwa',
       enforce: 'post',
-      transform(html) {
-        return injectServiceWorker(html, options)
+      apply: 'build',
+      configResolved(config) {
+        viteConfig = config
+        options = resolveOptions(userOptions, viteConfig)
       },
-    },
-    generateBundle(_, bundle) {
-      bundle[FILE_MANIFEST] = {
-        isAsset: true,
-        type: 'asset',
-        name: undefined,
-        source: `${JSON.stringify(options.manifest, null, options.minify ? 0 : 2)}\n`,
-        fileName: FILE_MANIFEST,
-      }
-      if (options.injectRegister === 'import' && !existsSync(join(viteConfig.root, 'public', FILE_SW_REGISTER))) {
-        bundle[FILE_SW_REGISTER] = {
+      transformIndexHtml: {
+        enforce: 'post',
+        transform(html) {
+          return injectServiceWorker(html, options)
+        },
+      },
+      generateBundle(_, bundle) {
+        bundle[FILE_MANIFEST] = {
           isAsset: true,
           type: 'asset',
           name: undefined,
-          source: generateSWRegister(options),
-          fileName: FILE_SW_REGISTER,
+          source: `${JSON.stringify(options.manifest, null, options.minify ? 0 : 2)}\n`,
+          fileName: FILE_MANIFEST,
         }
-      }
+
+        // if virtual register is requested, do not inject.
+        if (options.injectRegister === 'auto')
+          options.injectRegister = useRegister ? null : 'import'
+
+        if (options.injectRegister === 'import' && !existsSync(join(viteConfig.root, 'public', FILE_SW_REGISTER))) {
+          bundle[FILE_SW_REGISTER] = {
+            isAsset: true,
+            type: 'asset',
+            name: undefined,
+            source: generateSimpleSWRegister(options),
+            fileName: FILE_SW_REGISTER,
+          }
+        }
+      },
+      async closeBundle() {
+        if (!viteConfig.build.ssr) {
+          if (options.strategies === 'injectManifest')
+            await injectManifest(options.injectManifest)
+          else
+            await generateSW(options.workbox)
+        }
+      },
     },
-    async closeBundle() {
-      if (!viteConfig.build.ssr) {
-        if (options.strategies === 'injectManifest')
-          await injectManifest(options.injectManifest)
-        else
-          await generateSW(options.workbox)
-      }
+    {
+      name: 'vite-plugin-pwa:virtual',
+      configResolved(config) {
+        viteConfig = config
+        options = resolveOptions(userOptions, viteConfig)
+      },
+      resolveId(id, importer, _, ssr) {
+        if (ssr || options.injectRegister !== 'auto')
```

Comment:
When `registerType: 'autoUpdate'` don't generate the virtual module. 

The if statement should be:
```javascript
if (ssr || options.injectRegister !== 'auto' || options.registerType === 'autoUpdate')
```

Reply:
forgot it, the virtual is used to notify offline ready... and is protected to avoid refresh